### PR TITLE
Support arb number store objects

### PR DESCRIPTION
**Pull request by ssaucedo**

### CODE REVIEW

File: boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java [hunk 1]
```diff
@@ -66,21 +66,37 @@ public void run(BossConfiguration config, Environment env) throws Exception {
...
         SwaggerConfiguration swagger = config.getSwaggerConfiguration();
         // Set up the resources themselves.
         env.jersey().register(new ObjectResource(gBossAPI));
         env.jersey().register(new AllObjectsResource(gBossAPI));
         setSwaggerConfiguration(config, env, swagger);
-        if ( localConf.type == ObjectStoreType.FCS || cloudConf.type == ObjectStoreType.FCS ) {
-            env.jersey().register(new FCSResource());
-        }
+
+        for (Map.Entry<String, ObjectStoreConfiguration> entry : objectStoreConfigurationMap.entrySet()) {
```
Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:77]: How about this, which is a little less verbose:
for ( ObjectStoreConfiguration conf : objectStoreConfigurationMap.values() )
  if ( conf.type == ObjectStoreType.FCS )
    env.jersey().register(new FCSResource());

Comment (vvicario) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:77]: In the new code, it's similar.  We check if there exists some object store with type FCS, and if it exists, we register an FCSResource, I'm noticed that we never break the loop. In order to register one, and only one resource, we should add a break clause.

Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:77]: Yes, Veronica.  Absolutely true.  Should be a break.

Comment (vvicario) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:77]: I'll add it. Thanks

File: boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java [hunk 2]
```diff
@@ -66,21 +66,37 @@ public void run(BossConfiguration config, Environment env) throws Exception {
...
+        	if (entry.getValue().type == ObjectStoreType.FCS){
+    			env.jersey().register(new FCSResource());
+    		}
+    	}
+    	
+        
     }
 
+    protected static Map<String, ObjectStore> getObjectStoresMap(Map<String,ObjectStoreConfiguration> objectStoreConfigurationMap) throws Exception {
```
Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java]: Let's make this private -- we're not planning to extend BossApplication.

File: boss-ws/src/main/java/org/genomebridge/boss/http/BossConfiguration.java [hunk 3]
```diff
@@ -19,33 +22,26 @@ public DataSourceFactory getDataSourceFactory() {
...
     @JsonProperty
-    private ObjectStoreConfiguration cloudStore = new ObjectStoreConfiguration();
+    private Map<String,ObjectStoreConfiguration> objectStores = new  HashMap<String,ObjectStoreConfiguration>();
 
     @Valid
     @NotNull
     @JsonProperty
     private SwaggerConfiguration swagger = new SwaggerConfiguration();
+
+	public Map<String, ObjectStoreConfiguration> getObjectStores() {
```
Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossConfiguration.java]: It would be prettier to move this method up with the other public methods.

**Comment by tedsharpe** [comment]
The StoragePlatform enum should be eliminated -- it no longer describes the reality where valid values are not enumerated, but come from the config file.  It can be replaced by defining some string constants:
for "opaqueURI" in DatabaseBossAPI
and for "opaqueURI" and "mockStore" in the unit tests

File: boss-ws/src/test/resources/boss-config.yml [hunk 4]
```diff
@@ -11,19 +11,22 @@ database:
...
-  endpoint: http://localhost:8180/fcs
-  type: FCS
-localStore:
-  username: ecsUserName
-  password: ecsPassword
-  bucket: bossbucket
-  endpoint: https://ecsserver.broadinstitute.org
-  pathStyleAccess: true
-  type: S3
+objectStores:
```
Comment (tedsharpe) [inline boss-ws/src/test/resources/boss-config.yml:14]: We really only need one object store for unit testing -- we could call it "mockStore".  The unit tests will need to be changed to use this.

File: boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java [hunk 5]
```diff
@@ -248,12 +258,20 @@ else if ( req.httpMethod.equals(HttpMethod.GET) ||
         else {
             long timeout = now.getTime() + 1000L*req.validityPeriodSeconds;
             ObjectStore objStore = getObjectStore(rec.storagePlatform);
+
+            if (objStore != null && objStore.isReadOnly()){
+                if ( req.httpMethod.equals(HttpMethod.PUT)
+                        || req.httpMethod.equals(HttpMethod.HEAD ))
```
Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java]: HEAD method is OK on a read-only store.

**Comment by tedsharpe** [comment]
We need a new unit test for the read-only functionality.

File: boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java [hunk 6]
```diff
@@ -296,23 +318,26 @@ public ErrorDesc getResumableUploadURL(String objectId, String userName, CopyRes
...
         resp.uri = objStore.generateResumableUploadURL(rec.objectName);
         return error;
-
     }
-
     private BossDAO getDao() {
         return mDBI.onDemand(BossDAO.class);
     }
 
     private ObjectStore getObjectStore( String storagePlatform ) {
```
Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java]: This method is no longer useful.  It would be better to inline it in all the places it's called.  Then it will be easy to return an ErrorDesc for an invalid storagePlatform.

**Comment by tedsharpe** [comment]
That's all that I see right now, Santiago.  I'll wait for your changes.

**Comment by vvicario** [comment]
@tedsharpe it's ready for review.

**Comment by tedsharpe** [comment]
This is almost ready to go.  There are two things we need to do:

1) For unit testing we have only three valid object store types that we can test -- opaqueURI, mockStore, and mockStoreReadOnly.  Eliminate CLOUD_STORE, and fix the tests that use CLOUD_STORE to use the mockStore instead.  Change boss-config.yml to define mockStore and mockStoreReadOnly -- remove localStore and cloudStore.

2) There is a possibility of NullPointerExceptions in DatabaseBossAPI.  We need a new method like this:
    private ErrorDesc badStore(String storagePlatform) {
      return new ErrorDesc(Response.Status.INTERNAL_SERVER_ERROR,
                                 String.format(getMessage("funkyStoragePlatform"),storagePlatform));
    }
The right status code is INTERNAL_SERVER_ERROR because the bad value came from the database, not from the user, and this is our misconfiguration:  a storagePlatform that was once valid is no longer configured. 
And then, you need to test for null after mObjectStore.get and call this new method if it returns null.  Typically it will look something like this:
    ObjectStore store = mObjectStore.get(rec.storagePlatform);
    if ( store == null ) return badStore(rec.storagePlatform);
    if ( store.isReadOnly() ) return readOnlyStoreErr("readOnlyStore");
Unfortunately, each case where we call mObjectStore.get is slightly different, and will need to be analyzed separately.  It's OK to get null when the storagePlatform is opaqueURI -- so that needs to be tested first.  For example, in resolveObjectForCopying and getResumableUploadURL we've already rejected the opaqueURI case, so the code above will work fine.  However, in deleteObject we haven't tested for the opaqueURI case, so we'll have to do that first.  Something like:
    ObjectStore store = null;
    if ( !OPAQUEURI.equals(rec.storagePlatform) ) {
        if ( (store = mObjectStore.get(rec.storagePlatform)) == null )
            return badStore(rec.storagePlatform);
        if ( store.isReadOnly() )
            return readOnlyStoreErr("readOnlyStore");
    }
The insertObject case is fine as it stands, because we validate the storagePlatform in testCreationValidity.

The code suggestions I typed in haven't been compiled, and your code doesn't have to look exactly the same, anyway.  I was just trying to communicate clearly.

**Comment by tedsharpe** [comment]
In DatabaseBossAPI before line 264 we need
if ( objStore==null ) return badStoreErr(rec.storagePlatform);

Other than that it looks good.
Squash the commits, but don't merge.  Before we merge into develop we'll have to have Ray Pete make changes to the deployed configuration files to accommodate the new readOnly attribute for object stores.

Thanks for your work.
